README
NAME
    Hailo - A pluggable Markov engine analogous to MegaHAL

SYNOPSIS
    This is the synopsis for using Hailo as a module. See hailo for
    command-line invocation.

     # Hailo requires Perl 5.10
     use 5.010;
     use strict;
     use warnings;
     use Hailo;

     # Construct a new in-memory Hailo using the SQLite backend. See
     # backend documentation for other options.
     my $hailo = Hailo->new;

     # Various ways to learn
     my @train_this = qw< I like big butts and I can not lie >;
     $hailo->learn(\@train_this);
     $hailo->learn($_) for @train_this;

     # Heavy-duty training interface. Backends may drop some safety
     # features like journals or synchronous IO to train faster using
     # this mode.
     $hailo->learn("megahal.trn");
     $hailo->learn($filehandle);

     # Make the brain babble
     say $hailo->reply("hello good sir.");

DESCRIPTION
    Hailo is a fast and lightweight markov engine intended to replace
    AI::MegaHAL. It has a Mouse (or Moose) based core with pluggable storage
    and tokenizer backends.

    It is similar to MegaHAL in functionality, the main differences (with
    the default backends) being better scalability, drastically less memory
    usage, an improved tokenizer, and tidier output.

    With this distribution, you can create, modify, and query Hailo brains.
    To use Hailo in event-driven POE applications, you can use the
    POE::Component::Hailo wrapper. One example is
    POE::Component::IRC::Plugin::Hailo, which implements an IRC chat bot.

  Etymology
    *Hailo* is a portmanteau of *HAL* (as in MegaHAL) and failo
    <http://identi.ca/failo>.

Backends
    Hailo supports pluggable storage and tokenizer backends, it also
    supports a pluggable UI backend which is used by the hailo command-line
    utility.

  Storage
    Hailo can currently store its data in either a SQLite, PostgreSQL or
    MySQL database, more backends were supported in earlier versions but
    they were removed as they had no redeeming quality.

    SQLite is the primary target for Hailo. It's much faster and uses less
    resources than the other two. It's highly recommended that you use it.

    This benchmark shows how the backends compare when training on the small
    testsuite dataset as reported by the utils/hailo-benchmark utility
    (found in the distribution):

                             Rate DBD::Pg DBD::mysql DBD::SQLite/file DBD::SQLite/memory
        DBD::Pg            2.22/s      --       -33%             -49%               -56%
        DBD::mysql         3.33/s     50%         --             -23%               -33%
        DBD::SQLite/file   4.35/s     96%        30%               --               -13%
        DBD::SQLite/memory 5.00/s    125%        50%              15%                 --

    Under real-world workloads SQLite is much faster than these results
    indicate since the time it takes to train/reply is relative to the
    existing database size. Here's how long it took to train on a 214,710
    line IRC log on a Linode 1080 with Hailo 0.18:

    *   SQLite

         real    8m38.285s
         user    8m30.831s
         sys     0m1.175s

    *   MySQL

         real    48m30.334s
         user    8m25.414s
         sys     4m38.175s

    *   PostgreSQL

         real    216m38.906s
         user    11m13.474s
         sys     4m35.509s

    In the case of PostgreSQL it's actually much faster to first train with
    SQLite, dump that database and then import it with psql(1), see failo's
    README <http://github.com/hinrik/failo> for how to do that.

    However when replying with an existing database (using
    utils/hailo-benchmark-replies) yields different results. SQLite can
    reply really quickly without being warmed up (which is the typical
    usecase for chatbots) but once PostgreSQL and MySQL are warmed up they
    start replying faster:

    Here's a comparison of doing 10 replies:

                            Rate PostgreSQL MySQL SQLite-file SQLite-file-28MB SQLite-memory
        PostgreSQL        71.4/s         --  -14%        -14%             -29%          -50%
        MySQL             83.3/s        17%    --          0%             -17%          -42%
        SQLite-file       83.3/s        17%    0%          --             -17%          -42%
        SQLite-file-28MB 100.0/s        40%   20%         20%               --          -30%
        SQLite-memory      143/s       100%   71%         71%              43%            --

    In this test MySQL uses around 28MB of memory (using Debian's
    my-small.cnf) and PostgreSQL around 34MB. Plain SQLite uses 2MB of cache
    but it's also tested with 28MB of cache as well as with the entire
    database in memory.

    But doing 10,000 replies is very different:

                           Rate SQLite-file PostgreSQL SQLite-file-28MB MySQL SQLite-memory
        SQLite-file      85.1/s          --        -7%             -18%  -27%          -38%
        PostgreSQL       91.4/s          7%         --             -12%  -21%          -33%
        SQLite-file-28MB  103/s         21%        13%               --  -11%          -25%
        MySQL             116/s         37%        27%              13%    --          -15%
        SQLite-memory     137/s         61%        50%              33%   18%            --

    Once MySQL gets more memory (using Debian's my-large.cnf) and a chance
    to warm it starts yielding better results (I couldn't find out how to
    make PostgreSQL take as much memory as it wanted):

                       Rate         MySQL SQLite-memory
        MySQL         121/s            --          -12%
        SQLite-memory 138/s           14%            --

  Tokenizer
    By default Hailo will use the word tokenizer to split up input by
    whitespace, taking into account things like quotes, sentence terminators
    and more.

    There's also a the character tokenizer. It's not generally useful for a
    conversation bot but can be used to e.g. generate new words given a list
    of existing words.

UPGRADING
    Hailo makes no promises about brains generated with earlier versions
    being compatable with future version and due to the way Hailo works
    there's no practical way to make that promise.

    If you're maintaining a Hailo brain that you want to keep using you
    should save the input you trained it on and re-train when you upgrade.

    The reason for not offering a database schema upgrade for Hailo is
    twofold:

    *   We're too lazy to maintain database upgrade scripts for every
        version.

    *   Even if we weren't there's no way to do it right.

    The reason it can't be done right is that Hailo is always going to lose
    information present in the input you give it. How input tokens get split
    up and saved to the storage backend depends on the version of the
    tokenizer being used and how that input gets saved to the database.

    For instance if an earlier version of Hailo tokenized "foo+bar" simply
    as "foo+bar" but a later version split that up into "foo", "+", "bar",
    then an input of ""foo+bar are my favorite metasyntactic variables""
    wouldn't take into account the existing "foo+bar" string in the
    database.

    Tokenizer changes like this would cause the brains to accumulate garbage
    and would leave other parts in a state they wouldn't otherwise have
    gotten into. There have been similar changes to the database format
    itself.

    In short, learning is lossy so an accurate conversion is impossible.

ATTRIBUTES
  "brain_resource"
    The name of the resource (file name, database name) to use as storage.
    There is no default. Whether this gets used at all depends on the
    storage backend, currently only SQLite uses it.

  "save_on_exit"
    A boolean value indicating whether Hailo should save its state before
    its object gets destroyed. This defaults to true and will simply call
    save at "DEMOLISH" time.

  "order"
    The Markov order (chain length) you want to use for an empty brain. The
    default is 2.

  "storage_class"
    The storage backend to use. Default: 'SQLite'.

  "tokenizer_class"
    The tokenizer to use. Default: 'Words';

  "ui_class"
    The UI to use. Default: 'ReadLine';

  "storage_args"
  "tokenizer_args"
  "ui_args"
    A "HashRef" of arguments for storage/tokenizer/ui backends. See the
    documentation for the backends for what sort of arguments they accept.

METHODS
  "new"
    This is the constructor. It accepts the attributes specified in
    "ATTRIBUTES".

  "learn"
    Takes a string or an array reference of strings and learns from them.

  "train"
    Takes a filename, filehandle or array reference and learns from all its
    lines. If a filename is passed, the file is assumed to be UTF-8 encoded.
    Unlike "learn", this method sacrifices some safety (disables the
    database journal, fsyncs, etc) for speed while learning.

  "reply"
    Takes an optional line of text and generates a reply that might be
    relevant.

  "learn_reply"
    Takes a string argument, learns from it, and generates a reply that
    might be relevant. This is equivalent to calling learn followed by
    reply.

  "save"
    Tells the underlying storage backend to save its state, any arguments to
    this method will be passed as-is to the backend.

  "stats"
    Takes no arguments. Returns the number of tokens, expressions, previous
    token links and next token links.

PRIVATE METHODS
  "run"
    Run Hailo in accordance with the the attributes that were passed to it,
    this method is called by the hailo command-line utility and the Hailo
    test suite, it's behavior is subject to change.

SUPPORT
    You can join the IRC channel *#hailo* on FreeNode if you have questions.

BUGS
    Bugs, feature requests and other issues are tracked in Hailo's issue
    tracker on Github <http://github.com/hinrik/hailo/issues>.

SEE ALSO
    *   Hailo::UI::Web - A Catalyst and jQuery powered web interface to
        Hailo

    *   POE::Component::Hailo - A non-blocking POE wrapper around Hailo

    *   POE::Component::IRC::Plugin::Hailo - A Hailo IRC bot plugin

    *   <http://github.com/hinrik/failo> - Failo, an IRC bot that uses Hailo

LINKS
    *   <http://bit.ly/hailo_rewrite_of_megahal> - Hailo: A Perl rewrite of
        MegaHAL, A blog posting about the motivation behind Hailo

AUTHORS
    Hinrik Örn Sigurðsson, hinrik.sig@gmail.com

    Ævar Arnfjörð Bjarmason <avar@cpan.org>

LICENSE AND COPYRIGHT
    Copyright 2010 Hinrik Örn Sigurðsson and Ævar Arnfjörð Bjarmason
    <avar@cpan.org>

    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Hailo

{{$NEXT}}

0.70 2011-09-18 19:15:17

    - When using --train-fast, remove the "flushing cache" message when done

    - Word tokenizer:
        * Improve tokenization of email addresses
        * Use backspace instead of escape as a magic character when
          capitalizing text in multiple passes, since it's less likely to
          appear in tokens.
        * Preserve casing of words like "ATMs"

0.69 2011-05-07 04:02:38

    - Scored engine: Prefer shorter replies, like MegaHAL/cobe do

    - Word tokenizer:
        * Improve matching/capitalization of filenames and domain names
        * Match timestamps as single tokens
        * Match IRC nicks (<foobar>, <@foobar>, etc) as single tokens
        * Match IRC channel names (#foo, &bar, +baz)
        * Match various prefixes and postfixes with numbers
        * Match "#1" and "#1234" as single tokens
        * Match </foo> as a single token

    - Depend on MouseX::Getopt 0.33 to fix test failures

0.68 2011-05-03 13:16:05

    - Speed up the learning of repetitive sentences by caching more

    - Added Hailo::Engine::Scored, which generates multiple replies (limited
      by time or number of iterations) and returns the best one. Based on
      code from Peter Teichman's Cobe project.

    - Fixed a bug which caused the tokenizer to be very slow at capitalizing
      replies which contain things like "script/osm-to-tilenumbers.pl"

    - Speed up learning quite a bit (up to 25%) by using more efficient SQL.

    - Add --train-fast to speed up learning by up to an additional 45% on
      large brains by using aggressive caching. This uses a lot of memory.
      Almost 600MB with SQLite on a 64bit machine for a brain which
      eventually takes 134MB on disk (trained from a 350k line IRC log).

    - Word tokenizer:
        * Preserve casing of Emacs key sequences like "C-u"
        * Don't capitalize words after ellipses (e.g. "Wait... what?")
        * When adding a full stop to paragraphs which end with a quoted word,
          add it inside the quotes (e.g. "I heard him say 'hello there.'")
        * Make it work correctly when the input has newlines

0.67 2011-04-27 23:33:08

    - Word tokenizer: Improve matching of Perl class names

    - Require MooseX::StrictConstructor 0.16 for Moose 2.0 compatibility

0.66 2011-04-27 07:37:45

    - Die when the user tries --stats without a brain

    - Word tokenizer: stopped matching numbers separately in many cases.
      "95GB", "3D", "800x600" are all single tokens now. This should reduce
      the volume of nonsensical word-number pairs in the output.

0.65 2011-04-26 19:28:27

    - Expand "~" and such in the brain filename argument.

    - Word tokenizer:
        * Recognize "e.g." and other abbreviations as single tokens.
        * Recognize '–' as a word-separating dash.
        * Put a full stop after words separated by dots, like "sources.list".
        * Fix capitalization problems caused by "..."
        * Capitalize "foo!" and "foo."
        * Preserve casing of words such as "POE-Component-IRC"
        * Catch "Acme::POE::Tree" as one word, and preserve its casing
        * Catch "rollin'" as one word when it makes sense
        * Catch 'foo-" as one word when it makes sense, and capitalize at
          start of a sentence
        * Capitalize quoted words at the start of a line

    - The --help option didn't work with Moose >=1.25 (RT #67548). Fixed.

0.64 2010-12-10 11:09:08

    - Say "X lines of Y" instead of "X of Y" when training with
      hailo(1).

    - Bump README.pod in the distro to 0.59~1

0.63 2010-12-09 09:03:30
    - Mark the failing tests in t/command/shell.t as TODO for now.

0.62 2010-12-06 03:30:07
    - Really fix t/command/shell.t this time. I had the arguments to
      the test function in the wrong order.

0.61 2010-12-03 06:47:22

    - Use Class::Load instead of Class::MOP::load_class and eval'd require()

    - Fix rare test failure in t/command/shell.t

0.60 2010-11-09 01:35:49

    - Match email addresses and don't capitalize them

    - Don't insert additional spaces in some parsing edge-cases

0.59 2010-10-23 21:20:22

    - Word tokenizer: Match combining characters in words on Perl >=5.12

0.58 2010-10-22 03:34:08

    - Forgot to remove "use Text::Unidecode;" from Words.pm after dropping
      that dependency in the previous release.

    - Skip utf8-text.t on Perl <5.12 due to differences in Unicode matching.

0.57 2010-10-21 01:25:09

    - Fix typo in Schema.pm which messed up the type of the token.text
      SQL column (only relevant to non-SQLite DBs)

    - More robust matching of non-word immediately following URIs

0.56 2010-10-18 05:15:10

    - Add Test::Expect and Test::Synopsis to TestRequires

    - Use YAML::XS in utf8-text.t for clarity, and make the test
      optional.

0.55 2010-10-16 17:58:00

    - Scrap the "normal word" matching constraint from the previous release.
      It was failing to match words like "4.1GB", causing an infinite loop.
      Instead, we now assign different spacing attributes to components of
      conjoined words of varying types (e.g. recognizing that "4.1GB" is a
      normal word ("4.1") followed by a postfix ("GB").

    - Don't match '>' as part of the URI in "<http://google.com>".

0.54 2010-10-16 10:10:19

    - Fix the previously broken `save_on_exit' so that it does what
      it's supposed to do.

      The `save_on_exit' option in Hailo.pm never worked, and nor did
      hailo's corresponding --autosave option. Fixed that by
      correcting some method name confusion, and solved a BUG by
      finally adding a test for this option.

0.53 2010-10-15 21:29:02

    - hailo command: Fix off-by-one error when reporting the number of lines
      learned when training.

    - Don't truncate the brain file after loading it with in_memory enabled.

    - Word tokenizer improvements:
      * Recognize "´" as an apostrophe when matching words.

lib/Hailo/Storage.pm
package Hailo::Storage;

use 5.010;
use Any::Moose;
use Any::Moose 'X::StrictConstructor';
use DBI;
use Hailo::Storage::Schema;

has dbd => (
    isa           => 'Str',
    is            => 'ro',
    lazy_build    => 1,
    documentation => "The DBD::* driver we're using",
);

has dbd_options => (
    isa           => 'HashRef',
    is            => 'ro',
    lazy_build    => 1,
    documentation => 'Options passed as the last argument to DBI->connect()',
);

sub _build_dbd_options {
    my ($self) = @_;
    return {
        RaiseError => 1
    };
}

has dbh => (
    isa           => 'DBI::db',
    is            => 'ro',
    lazy_build    => 1,
    documentation => 'Our DBD object',
);

sub _build_dbh {
    my ($self) = @_;
    my $dbd_options = $self->dbi_options;

    return DBI->connect($self->dbi_options);
};

has dbi_options => (
    isa           => 'ArrayRef',
    is            => 'ro',
    auto_deref    => 1,
    lazy_build    => 1,
    documentation => 'Options passed to DBI->connect()',
);

sub _build_dbi_options {
    my ($self) = @_;
    my $dbd = $self->dbd;
    my $dbd_options = $self->dbd_options;
    my $db = $self->brain // '';

    my @options = (
        "dbi:$dbd:dbname=$db",
        '',
        '',
        $dbd_options,
    );

    return \@options;
}

has _engaged => (
    isa           => 'Bool',
    is            => 'rw',
    default       => 0,
    documentation => 'Have we done setup work to get this database going?',
);

has sth => (
    isa        => 'HashRef',
    is         => 'ro',
    lazy_build => 1,
    documentation => 'A HashRef of prepared DBI statement handles',
);

sub _build_sth {
    my ($self) = @_;
    return Hailo::Storage::Schema->sth($self->dbd, $self->dbh, $self->order);
}

has _boundary_token_id => (
    isa => 'Int',
    is  => 'rw',
);

# bootstrap the database
sub _engage {
    my ($self) = @_;

    if ($self->initialized) {
        # Check the order we've been given and retrieve it from the
        # database if there's nothing odd going on.
        $self->_engage_initialized_check_and_set_order;

        # Likewise for the Tokenizer
        $self->_engage_initialized_check_and_set_tokenizer;

        $self->sth->{token_id}->execute(0, '');
        my $id = $self->sth->{token_id}->fetchrow_array;
        $self->_boundary_token_id($id);
    }
    else {
        Hailo::Storage::Schema->deploy($self->dbd, $self->dbh, $self->order);

        # Set metadata in the database for use by subsequent
        # invocations
        {
            # Don't change order again
            my $order = $self->order;
            $self->sth->{set_info}->execute('markov_order', $order);

            # Warn if the tokenizer changes
            my $tokenizer = $self->tokenizer_class;
            $self->sth->{set_info}->execute('tokenizer_class', $tokenizer);
        }

        $self->sth->{add_token}->execute(0, '');
        $self->sth->{last_token_rowid}->execute();
        my $id = $self->sth->{last_token_rowid}->fetchrow_array();
        $self->_boundary_token_id($id);
    }

    $self->_engaged(1);

    return;
}

sub _engage_initialized_check_and_set_order {
    my ($self) = @_;

    my $sth = $self->dbh->prepare(qq[SELECT text FROM info WHERE attribute = ?;]);
    $sth->execute('markov_order');
    my $db_order = $sth->fetchrow_array();

    my $my_order = $self->order;
    if ($my_order != $db_order) {
        if ($self->hailo->{has_custom_order}->()) {
            die <<"DIE";
You've manually supplied an order of `$my_order' to Hailo but you're
loading a brain that has the order `$db_order'.

Hailo will automatically load the order from existing brains, however
you've constructed Hailo and manually specified an order not
equivalent to the existing order of the database.

Either supply the correct order or omit the order attribute
altogether. We could continue but I'd rather die since you're probably
expecting something I can't deliver.
DIE
        }

        $self->order($db_order);
        $self->hailo->{set_order}->($db_order);
    }

    return;
}

sub _engage_initialized_check_and_set_tokenizer {
    my ($self) = @_;

    my $sth = $self->dbh->prepare(qq[SELECT text FROM info WHERE attribute = ?;]);
    $sth->execute('tokenizer_class');
    my $db_tokenizer_class = $sth->fetchrow_array;
    my $my_tokenizer_class = $self->tokenizer_class;

    # defined() because we can't count on old brains having this
    if (defined $db_tokenizer_class
        and $my_tokenizer_class ne $db_tokenizer_class) {
        if ($self->hailo->{has_custom_tokenizer_class}->()) {
            die <<"DIE";
You've manually supplied a tokenizer class `$my_tokenizer_class' to
Hailo, but you're loading a brain that has the tokenizer class
`$db_tokenizer_class'.

Hailo will automatically load the tokenizer class from existing
brains, however you've constructed Hailo and manually specified an
tokenizer class not equivalent to the existing tokenizer class of the
database.

Either supply the correct tokenizer class or omit the order attribute
altogether. We could continue but I'd rather die since you're probably
expecting something I can't deliver.
DIE
        }

        $self->tokenizer_class($db_tokenizer_class);
        $self->hailo->{set_tokenizer_class}->($db_tokenizer_class);
    }

    return;
}

sub start_training {
    my ($self) = @_;
    $self->_engage() unless $self->_engaged;
    $self->start_learning();
    return;
}

sub stop_training {
    my ($self) = @_;
    $self->stop_learning();
    return;
}

sub start_learning {
    my ($self) = @_;
    $self->_engage() unless $self->_engaged;

    # start a transaction
    $self->dbh->begin_work;
    return;
}

sub stop_learning {
    my ($self) = @_;
    # finish a transaction
    $self->dbh->commit;
    return;
}

# See if SELECT count(*) FROM info; fails. If not we assume that we
# have an up and running database.
sub initialized {
    my ($self) = @_;
    my $dbh = $self->dbh;

    my ($err, $warn, $res);
    eval {
        # SQLite will warn 'no such table info'
        local $SIG{__WARN__} = sub { $err = $_[0] };

        # If it doesn't warn trust that it dies here
        local ($@, $!);
        $res = $dbh->do("SELECT count(*) FROM info;");
    };

    return (not $err and not $warn and defined $res);
}

# return some statistics
sub totals {
    my ($self) = @_;
    $self->_engage() unless $self->_engaged;

    $self->sth->{token_total}->execute();
    my $token = $self->sth->{token_total}->fetchrow_array - 1;
    $self->sth->{expr_total}->execute();
    my $expr = $self->sth->{expr_total}->fetchrow_array // 0;
    $self->sth->{prev_total}->execute();
    my $prev = $self->sth->{prev_total}->fetchrow_array // 0;
    $self->sth->{next_total}->execute();
    my $next = $self->sth->{next_total}->fetchrow_array // 0;

    return $token, $expr, $prev, $next;
}

__PACKAGE__->meta->make_immutable;

=encoding utf8

=head1 NAME

Hailo::Storage - A base class for L<Hailo> L<storage|Hailo::Role::Storage> backends

=head1 METHODS

The following two methods must to be implemented by subclasses:

=head2 C<_build_dbd>

Should return the name of the database driver (e.g. 'SQLite') which will be
passed to L<DBI|DBI>.

=head2 C<_build_dbd_options>

Subclasses can override this method to add options of their own. E.g:

    override _build_dbd_options => sub {
        return {
            %{ super() },
            sqlite_unicode => 1,
        };
    };

=head1 Comparison of backends

This benchmark shows how the backends compare when training on the
small testsuite dataset as reported by the F<utils/hailo-benchmark>
utility (found in the distribution):

                         Rate DBD::Pg DBD::mysql DBD::SQLite/file DBD::SQLite/memory
    DBD::Pg            2.22/s      --       -33%             -49%               -56%
    DBD::mysql         3.33/s     50%         --             -23%               -33%
    DBD::SQLite/file   4.35/s     96%        30%               --               -13%
    DBD::SQLite/memory 5.00/s    125%        50%              15%                 --

Under real-world workloads SQLite is much faster than these results
indicate since the time it takes to train/reply is relative to the
existing database size. Here's how long it took to train on a 214,710
line IRC log on a Linode 1080 with Hailo 0.18:

=over

=item * SQLite

    real    8m38.285s
    user    8m30.831s
    sys     0m1.175s

=item * MySQL

    real    48m30.334s
    user    8m25.414s
    sys     4m38.175s

=item * PostgreSQL

    real    216m38.906s
    user    11m13.474s
    sys     4m35.509s

=back

In the case of PostgreSQL it's actually much faster to first train
with SQLite, dump that database and then import it with L<psql(1)>,
see L<failo's README|http://github.com/hinrik/failo> for how to do
that.

However, replying with an existing database (using
F<utils/hailo-benchmark-replies>) yields different results. SQLite can
reply really quickly without being warmed up (which is the typical
usecase for chatbots) but once PostgreSQL and MySQL are warmed up they
start replying faster:

Here's a comparison of doing 10 replies:

                        Rate PostgreSQL MySQL SQLite-file SQLite-file-28MB SQLite-memory
    PostgreSQL        71.4/s         --  -14%        -14%             -29%          -50%
    MySQL             83.3/s        17%    --          0%             -17%          -42%
    SQLite-file       83.3/s        17%    0%          --             -17%          -42%
    SQLite-file-28MB 100.0/s        40%   20%         20%               --          -30%
    SQLite-memory      143/s       100%   71%         71%              43%            --

In this test MySQL uses around 28MB of memory (using Debian's
F<my-small.cnf>) and PostgreSQL around 34MB. Plain SQLite uses 2MB of
cache but it's also tested with 28MB of cache as well as with the
entire database in memory.

But doing 10,000 replies is very different:

                       Rate SQLite-file PostgreSQL SQLite-file-28MB MySQL SQLite-memory
    SQLite-file      85.1/s          --        -7%             -18%  -27%          -38%
    PostgreSQL       91.4/s          7%         --             -12%  -21%          -33%
    SQLite-file-28MB  103/s         21%        13%               --  -11%          -25%
    MySQL             116/s         37%        27%              13%    --          -15%
    SQLite-memory     137/s         61%        50%              33%   18%            --

Once MySQL gets more memory (using Debian's F<my-large.cnf>) and a
chance to warm it starts yielding better results (I couldn't find out
how to make PostgreSQL take as much memory as it wanted):

                   Rate         MySQL SQLite-memory
    MySQL         121/s            --          -12%
    SQLite-memory 138/s           14%            --

=head1 AUTHOR

E<AElig>var ArnfjE<ouml>rE<eth> Bjarmason <avar@cpan.org>

Hinrik E<Ouml>rn SigurE<eth>sson, hinrik.sig@gmail.com

=head1 LICENSE AND COPYRIGHT

Copyright 2010 E<AElig>var ArnfjE<ouml>rE<eth> Bjarmason and
Hinrik E<Ouml>rn SigurE<eth>sson

This program is free software, you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
